Jetpack Mobile: Threat Details - custom views

ashiagr opened the pull request:
Parent: #13703

This PR creates custom `ViewTypes`  for `ThreatContext` and `FileName` for "Threat Details" screen.

References: [Scan design iteration i5](https://jetpackmobile.files.wordpress.com/2020/12/ios-scan-i5.png)

![device-2021-01-04-131412](https://user-images.githubusercontent.com/1405144/103512507-d4f3c480-4e8e-11eb-89c5-0951949f248c.png)

### To test
Prerequisite: 
1. Make sure both Scan feature flag and MySiteImprovements flag are set to on in the App settings.
2. WP.com account with a site having scan capability and different threats (e.g. http://do.wpmt.co/jp-scan-daily/).

- Login to the app using above WP.com account and select the site on the My Site tab
- Notice that scan menu is displayed
- Click scan menu item followed by threat item
- Notice that `ThreatContext.lines `are shown correctly with a highlighted line
- `ViewType` for `ThreatFileName` can be tested by putting below line above [this line](https://github.com/wordpress-mobile/WordPress-Android/blob/512b5f7f1252efca1ce0f7692fb65b7f03cbc89d/WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsListItemsBuilder.kt#L20) in `ThreatDetailsListItemsBuilder`: 
`threatModel.fileName?.let { items.add(ThreatFileNameState(UiStringText(it))) }`

### Notes
1. Ignore all styling
2. Tests covered in parent PR #13703

### Merge Instructions

1. Make sure #13675 is merged
2. Remove the "Not Ready for Merge label"
3. Merge this PR

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Introduce Adapter, VH for ThreatContext lines

ashiagr committed:
Introduce VH for FileName (path)

ashiagr committed:
Merge branch 'issue/13326-threat-details-skeleton-and-blank-content-uistate' into issue/13326-threat-details-custom-views

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/common/ViewType.kt:
```diff
@@ -8,5 +8,7 @@ enum class ViewType(val id: Int) {
-    BACKUP_SUB_HEADER(7)
+    BACKUP_SUB_HEADER(7),
+    THREAT_CONTEXT_LINES(8),
+    FILE_NAME(9)
```
@ashiagr - Is FILE_NAME is exclusive to the scan feature and if so, should we prefix it as such like we did for the others (backup, threats)? wdyt?

ashiagr replied:
Good point. Renamed in aa33ca5.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/utils/PaddingBackgroundColorSpan.kt:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * Creates BackgroundColorSpan with padding
+ * Credits: https://medium.com/@tokudu/android-adding-padding-to-backgroundcolorspan-179ab4fae187
```
@ashiagr - I never noticed the "non" spacing before, but this is a good hack. Nice find and enhancement

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsListItemsBuilder.kt:
```diff
@@ -1,12 +1,44 @@
 
 @Reusable
 class ThreatDetailsListItemsBuilder @Inject constructor() {
     // TODO ashiagr to be implemented
```
Tiny nitpick - Looks like you implemented this, do you want to keep the TODO any longer?

ashiagr replied:
This is not fully implemented in this PR. I've already removed it in the parent PR #13703 [here](https://github.com/wordpress-mobile/WordPress-Android/pull/13703/files#diff-8eeccaef6fff93719bd7f5f0d90b3a6655f9becdb4d1f0a01a1b9b6f80606fdfR28-R29).

zwarm commented:
@ashiagr - So far the code looks good, but I need your help (or a point in the right direction) for 

> Select a site (http://do.wpmt.co/jp-scan-daily/) with scan capability and a threat with "malicious code pattern" on the My Site tab

I have no idea how to set that up. Help please. Thanks

ashiagr committed:
Rename FILE_NAME to THREAT_FILE_NAME

ashiagr commented:
Thanks @zwarm! I've addressed your comments so far.

> So far the code looks good, but I need your help (or a point in the right direction)
> I have no idea how to set that up. Help please. Thanks

Pinged you on Slack with instructions.

zwarm reviewed (approved):
LGTM! Highlighted as expected. Nicely done. 👍
